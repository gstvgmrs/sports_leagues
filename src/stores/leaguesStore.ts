import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import { leaguesService } from '@/services/leaguesService'

interface League {
  idLeague: string
  strLeague: string
  strLeagueAlternate?: string
  strSport?: string
  strBadge?: string
}

export const useLeaguesStore = defineStore('leaguesStore', () => {
  const leagues = ref<League[]>([])
  const loading = ref(false)
  const error = ref<string | null>(null)
  const loadingBadges = ref<Record<string, boolean>>({})
  const searchQuery = ref('')
  const selectedSport = ref('')

  async function fetchLeagues() {
    try {
      loading.value = true
      error.value = null
      leagues.value = await leaguesService.fetchLeagues()
    } catch (err: unknown) {
      error.value = err instanceof Error ? err.message : 'Error fetching leagues'
    } finally {
      loading.value = false
    }
  }

  async function fetchLeagueBadge(idLeague: string) {
    if (loadingBadges.value[idLeague]) return

    try {
      loadingBadges.value[idLeague] = true
      const badgeUrl = await leaguesService.fetchLeagueBadge(idLeague)

      if (badgeUrl) {
        const itemIndex = leagues.value.findIndex((item) => item.idLeague === idLeague)
        if (itemIndex !== -1) {
          leagues.value[itemIndex] = { ...leagues.value[itemIndex], strBadge: badgeUrl }
        }
      }
    } catch (err: unknown) {
      console.error('Error fetching badge:', err)
    } finally {
      loadingBadges.value[idLeague] = false
    }
  }

  function setSearchQuery(query: string) {
    searchQuery.value = query
  }

  function setSportFilter(sport: string) {
    selectedSport.value = sport
  }

  const filteredLeagues = computed(() => {
    let filtered = leagues.value

    if (searchQuery.value.trim()) {
      filtered = filtered.filter((league) =>
        league.strLeague?.toLowerCase().includes(searchQuery.value.toLowerCase()),
      )
    }

    if (selectedSport.value && selectedSport.value !== 'All Sports') {
      filtered = filtered.filter((league) => league.strSport === selectedSport.value)
    }

    return filtered
  })

  const sportOptions = computed(() => {
    const sports = [
      ...new Set(
        leagues.value
          .map((item) => item.strSport)
          .filter((sport): sport is string => Boolean(sport)),
      ),
    ]
    return ['All Sports', ...sports]
  })

  return {
    leagues,
    loading,
    error,
    searchQuery,
    selectedSport,
    filteredLeagues,
    sportOptions,
    loadingBadges,
    fetchLeagues,
    setSearchQuery,
    setSportFilter,
    fetchLeagueBadge,
  }
})
