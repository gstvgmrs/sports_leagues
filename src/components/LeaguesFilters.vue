<script setup lang="ts">
import { useLeaguesStore } from '@/stores/leaguesStore'
import { storeToRefs } from 'pinia'
import Dropdown from '@/components/base/Dropdown.vue'
import SearchInput from '@/components/base/SearchInput.vue'

const leaguesStore = useLeaguesStore()
const { sportOptions } = storeToRefs(leaguesStore)

const handleFilterChange = (selectedOption: string) => {
  leaguesStore.setSportFilter(selectedOption)
}

const handleSearch = (query: string) => {
  leaguesStore.setSearchQuery(query)
}
</script>

<template>
  <SearchInput placeholder="Search by name" @search="handleSearch" class="flex-1" />
  <Dropdown :options="sportOptions" default-value="All Sports" @change="handleFilterChange" />
</template>
