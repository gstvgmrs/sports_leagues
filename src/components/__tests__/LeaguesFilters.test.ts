import { describe, it, expect, vi, beforeEach } from 'vitest'
import { mount } from '@vue/test-utils'
import { createPinia, setActivePinia } from 'pinia'
import LeaguesFilters from '@/components/LeaguesFilters.vue'
import { useLeaguesStore } from '@/stores/leaguesStore'

vi.mock('@/services/leaguesService', () => ({
  leaguesService: {
    fetchLeagues: vi.fn().mockResolvedValue([]),
    fetchLeagueBadge: vi.fn(),
  },
}))

describe('LeaguesFilters', () => {
  beforeEach(() => {
    setActivePinia(createPinia())
    vi.clearAllMocks()
  })

  it('should match snapshot', () => {
    const wrapper = mount(LeaguesFilters)

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render sport options in dropdown', async () => {
    const wrapper = mount(LeaguesFilters)
    const store = useLeaguesStore()

    store.leagues = [
      { idLeague: '1', strLeague: 'Test League', strSport: 'Soccer' },
      { idLeague: '2', strLeague: 'Test League 2', strSport: 'Basketball' },
    ]

    await wrapper.vm.$nextTick()

    const dropdown = wrapper.find('button')
    await dropdown.trigger('click')

    expect(wrapper.text()).toContain('All Sports')
    expect(wrapper.text()).toContain('Soccer')
    expect(wrapper.text()).toContain('Basketball')
  })
})
