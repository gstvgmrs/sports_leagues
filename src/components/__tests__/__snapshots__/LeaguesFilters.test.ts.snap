// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LeaguesFilters > should match snapshot 1`] = `
"<div class="relative flex-1">
  <div class="relative"><input placeholder="Search by name" type="text" class="px-4 py-2 pr-10 text-sm font-medium text-gray-700 bg-white border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-transparent w-full">
    <div class="absolute inset-y-0 right-0 flex items-center pr-3 pointer-events-none"><svg class="w-4 h-4 text-gray-400" fill="none" stroke="currentColor" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
        <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z"></path>
      </svg>
      <!--v-if-->
    </div>
  </div>
</div>
<div class="relative"><button class="flex items-center gap-2 px-4 py-2 text-sm font-medium text-gray-700 bg-white border border-gray-300 rounded-md hover:bg-gray-50 focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-transparent">All Sports <svg class="w-4 h-4 transition-transform duration-200" fill="none" stroke="currentColor" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
      <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M19 9l-7 7-7-7"></path>
    </svg></button>
  <div class="absolute right-0 mt-1 w-40 bg-white border border-gray-300 rounded-md shadow-lg z-10" style="display: none;">
    <div class="py-1"><button class="block w-full px-4 py-2 text-sm text-left text-gray-700 hover:bg-gray-100 bg-blue-50 text-blue-700">All Sports</button></div>
  </div>
</div>"
`;
